connectd/connectd.c

```diff
@@ -973,8 +981,46 @@ static void try_connect_one_addr(struct connecting *connect)
 			af = AF_INET6;
 			break;
 		case ADDR_TYPE_DNS:
-			// TODO: resolve with getaddrinfo and set af
-			break;
+			/* Resolve with getaddrinfo */
+			memset(&hints, 0, sizeof(hints));
+			hints.ai_socktype = SOCK_STREAM;
+			hints.ai_family = AF_UNSPEC;
+			hints.ai_protocol = 0;
+			hints.ai_flags = AI_ADDRCONFIG;
+			gai_err = getaddrinfo((char *)addr->u.wireaddr.addr,
+					      tal_fmt(tmpctx, "%d",
+						      addr->u.wireaddr.port),
+					      &hints, &ais);
+			if (gai_err != 0)
+				status_failed(STATUS_FAIL_INTERNAL_ERROR,
+					      "DNS with getaddrinfo gave: %s",
+					      gai_strerror(gai_err));
+			/* create new addrhints on-the-fly per result ... */
+			for (aii = ais; aii; aii = aii->ai_next) {
+				addrhint.itype = ADDR_INTERNAL_WIREADDR;
+				if (aii->ai_family == AF_INET) {
+					sa4 = (struct sockaddr_in *) aii->ai_addr;
+					wireaddr_from_ipv4(&addrhint.u.wireaddr,
+							   &sa4->sin_addr,
+							   addr->u.wireaddr.port);
+				} else if (aii->ai_family == AF_INET6) {
+					sa6 = (struct sockaddr_in6 *) aii->ai_addr;
+					wireaddr_from_ipv6(&addrhint.u.wireaddr,
+							   &sa6->sin6_addr,
+							   addr->u.wireaddr.port);
+				} else {
+					/* skip unsupported ai_family */
+					continue;
+				}
+				tal_arr_expand(&connect->addrs, addrhint);
+				/* fixes VALGRIND complains for next iteration */
+				addr = &connect->addrs[connect->addrnum];
```

Comment:
Gah, no, since you are now altering and potentially moving connect->addrs, valgrind is alerting you that addr can be stale!  But it's a real bug!

Reply:
(Better might be to save portnum in a var, and never reference addr in this loop)

Reply:
Reseating `addr` here should fix the bug, but yes, it is a real bug, not a Valgrind weirdness.  Nevertheless, doing this is still risky especially if the code is modified later, so it is better to just avoid it and save portnum in a new var.

Reply:
@rustyrussell @ZmnSCPxj 
Okay, I set a correct comment. I didn't get what you mean with "portnum" since this "portnum" is not even used in this file.

Reply:
We were suggesting keeping the port number in a local var, so you don't have to keep addr up-to-date in the loop (you access addr->u.wireaddr.port).  But it doesn't really help, since addr is used after the loop anyway...

Reply:
Ah, that's why I didn't understand it.
I thought about you meaning the `wireaddr.port` but it didn't make sense because addr is used for other stuff as well so I dismissed it.